src/MauiReactor/Internals/ServiceCollectionProvider.cs

```diff
@@ -1,16 +1,18 @@
 ﻿using MauiReactor.HotReload;
+using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.DependencyInjection.Extensions;
 using System.Reflection;
 
 namespace MauiReactor.Internals;
 
 internal static class ServiceCollectionProvider
 {
+    static readonly AsyncLocal<IServiceProvider?> _testingServiceProvider = new();
     static IServiceProvider? _serviceProvider = null!;
 
     public static IServiceProvider? ServiceProvider
-    { 
-        get => _serviceProvider;
+    {
+        get => _testingServiceProvider.Value ?? _serviceProvider;

```

Comment:
doing the async local null check here so that we do not break any public contracts, by using the service context in tests, all the threads will have an isolated service provider so will not stomp on each other. It is the same mechanism that ASP.Net Core uses for the request context.